ow_system_plugins/admin/controllers/logs.php

```diff
@@ -0,0 +1,162 @@
+ * The Initial Developer of the Original Code is Oxwall Foundation (http://www.oxwall.org/foundation).
+ * All portions of the code written by Oxwall Foundation are Copyright (c) 2011. All Rights Reserved.
+
+ * EXHIBIT B. Attribution Information
+ * Attribution Copyright Notice: Copyright 2011 Oxwall Foundation. All rights reserved.
+ * Attribution Phrase (not exceeding 10 words): Powered by Oxwall community software
+ * Attribution URL: http://www.oxwall.org/
+ * Graphic Image as provided in the Covered Code.
+ * Display of Attribution Information is required in Larger Works which are defined in the CPAL as a work
+ * which combines Covered Code or portions thereof with code not governed by the terms of the CPAL.
+ */
+
+/**
+ * Logs controller.
+ */
+class ADMIN_CTRL_Logs extends ADMIN_CTRL_Abstract
+{
+    const ENTRIES_PER_PAGE = 50;
+
+    /**
+     * @var BOL_LogService
+     */
+    protected $service;
+
+    public function __construct()
+    {
+        parent::__construct();
+
+        $this->service = BOL_LogService::getInstance();
+    }
+
+    /**
+     * Display list of all log entries.
+     */
+    public function index()
+    {
+        $pageParam = (int) ($_GET['page'] ?? null);
+        $typeParam = $_GET['type'] ?? null;
+        $keyParam = $_GET['key'] ?? null;
+        $searchParam = $_GET['search'] ?? null;
+
+        $page = $pageParam > 0 ? $pageParam - 1 : 0;
+
+        $first = self::ENTRIES_PER_PAGE * $page;
+
+        if ( !empty($typeParam) )
+        {
+            $entries = $this->service->findByTypePaginated($typeParam, $first, self::ENTRIES_PER_PAGE);
+            $totalEntries = $this->service->countByType($typeParam);
+        }
+        elseif ( !empty($keyParam) )
+        {
+            $entries = $this->service->findByKeyPaginated($keyParam, $first, self::ENTRIES_PER_PAGE);
+            $totalEntries = $this->service->countByKey($keyParam);
+        }
+        elseif ( !empty($searchParam) )
+        {
+            $entries = $this->service->findByQueryPaginated($searchParam, $first, self::ENTRIES_PER_PAGE);
+            $totalEntries = $this->service->countByQuery($searchParam);
+        }
+        else
+        {
+            $entries = $this->service->findAll($first, self::ENTRIES_PER_PAGE);
+            $totalEntries = $this->service->countAll();
+        }
+
+        $totalPages = floor($totalEntries / self::ENTRIES_PER_PAGE);
+
+        if ( empty($entries) )
+        {
+            $this->assign('isEmpty', true);
+            return;
+        }
+
+        $processedEntries = $this->service->processEntries($entries);
+
+        if ( $typeParam || $keyParam || $searchParam )
+        {
+            $this->assign('isFiltersApplied', true);
+            $this->assign('dropFiltersUrl', OW::getRouter()->urlForRoute('admin_developer_tools_logs'));
+        }
+
+        $paging = new BASE_CMP_Paging($page + 1, $totalPages, 10);
+        $this->assign('paging', $paging->render());
```

Comment:
@philchekhov you can also replace by `$this->addComponent('paging', $paging);`, but this is also acceptable.